core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -1003,9 +1003,14 @@ class LogManager(logDirs: Seq[File],
   /**
    * Map of log dir to logs by topic and partitions in that dir
    */
-  private def logsByDir: Map[String, Map[TopicPartition, Log]] = {
-    (this.currentLogs.toList ++ this.futureLogs.toList).toMap
-      .groupBy { case (_, log) => log.parentDir }
+  def logsByDir: Map[String, Map[TopicPartition, Log]] = {
+    val byDir = new mutable.HashMap[String, mutable.HashMap[TopicPartition, Log]]()
+    def addToDir(tp: TopicPartition, log: Log): Unit = {
+      byDir.getOrElseUpdate(log.parentDir, new mutable.HashMap[TopicPartition, Log]()).put(tp, log)
```

Comment:
In the other PR we used `AnyRefMap`. Any reason why we are not doing that here?

Reply:
We should also add a comment about why we are writing non idiomatic code here to avoid reverts.

Reply:
Good point. I will try it and retest and add a comment to prevent regression.

Reply:
I have addressed this and it's ready for another review pass.